File: src/rust/src/normalizers.rs
use extendr_api::prelude::*;
use tokenizers as tk;
#[extendr]
pub struct RNormalizer(pub tk::NormalizerWrapper);

#[extendr]
impl RNormalizer {
    pub fn new(normalizer: Robj) -> extendr_api::Result<Self> {
        if normalizer.inherits("RNormalizerNFC") {
            Ok(RNormalizer(
                <&RNormalizerNFC>::try_from(&normalizer)?.0.clone().into()
            ))
        } else if normalizer.inherits("RNormalizerNFKC") {
            Ok(RNormalizer(
                <&RNormalizerNFKC>::try_from(&normalizer)?.0.clone().into()
            ))
        } else {
            Err(Error::EvalError("Unsupported normalizer".into()))
        }
    }
}

#[extendr]
struct RNormalizerNFC(tk::normalizers::NFC);

#[extendr]
impl RNormalizerNFC {
    pub fn new() -> Self {
        RNormalizerNFC(tk::normalizers::NFC::default())
    }
}

#[extendr]
struct RNormalizerNFKC(tk::normalizers::NFKC);

#[extendr]
impl RNormalizerNFKC {
    pub fn new() -> Self {
        RNormalizerNFKC(tk::normalizers::NFKC::default())
    }
}

extendr_module! {
    mod normalizers;
    impl RNormalizer;
    impl RNormalizerNFC;
    impl RNormalizerNFKC;
}

File: R/normalizers.R
#' Generic class for normalizers
#' @export
#' @family normalizers
#' @export
tok_normalizer <- R6::R6Class(
  "tok_normalizer",
  public = list(
    #' @field .normalizer Internal pointer to normalizer object
    .normalizer = NULL,
    #' @description
    #' Initializes a tokenizer
    #' @param normalizer a raw pointer to a tokenizer
    initialize = function(normalizer) {
      if (inherits(normalizer, "RNormalizer")) {
        self$.normalizer <- normalizer
      } else {
        self$.normalizer <- RNormalizer$new(normalizer)  
      }
    }
  )
)

#' NFC normalizer
#' @export
#' @family normalizers
normalizer_nfc <- R6::R6Class(
  "tok_normalizer_nfc",
  inherit = tok_normalizer,
  public = list(
    #' @description
    #' Initializes the NFC normalizer
    initialize = function() {
      super$initialize(RNormalizerNFC$new())
    }  
  )
)

#' NFKC normalizer
#' @export
#' @family normalizers
normalizer_nfkc <- R6::R6Class(
  "tok_normalizer_nfkc",
  inherit = tok_normalizer,
  public = list(
    #' @description
    #' Initializes the NFKC normalizer
    initialize = function() {
      super$initialize(RNormalizerNFKC$new())
    }  
  )
)

